cmd/enterprise-portal/internal/dotcomdb/dotcomdb.go

```diff
@@ -1,32 +1,250 @@
+		FROM product_licenses
+		-- Get most recently created license key as the "active license"
+		ORDER BY product_subscription_id, created_at DESC
+	) active_license ON active_license.product_subscription_id = ps.id
+	LEFT JOIN (
+		SELECT
+			product_subscription_id,
+			digest(license_key, 'sha256') as license_key_hash
+		FROM product_licenses
+		WHERE access_token_enabled IS TRUE
+	) tokens ON tokens.product_subscription_id = ps.id
+`
+
+type GetCodyGatewayAccessAttributesOpts struct {
+	BySubscription *string
+	ByAccessToken  *string
+}
+
+func (r *Reader) GetCodyGatewayAccessAttributesBySubscription(ctx context.Context, subscriptionID string) (*CodyGatewayAccessAttributes, error) {
+	query := selectCodyGatewayAccessAttributes +
+		"WHERE ps.id = $1 GROUP BY ps.id"
+	row := r.conn.QueryRow(ctx, query, subscriptionID)
+	return r.scanOneCodyGatewayAccessAttributes(row)
+}
+
+func (r *Reader) GetCodyGatewayAccessAttributesByAccessToken(ctx context.Context, token string) (*CodyGatewayAccessAttributes, error) {
+	// Below is copy-pasta from 'func (t dbTokens) LookupProductSubscriptionIDByAccessToken'
+	// in 'cmd/frontend/internal/dotcom/productsubscription'.
+	if !strings.HasPrefix(token, productsubscription.AccessTokenPrefix) &&
+		!strings.HasPrefix(token, license.LicenseKeyBasedAccessTokenPrefix) {
+		return nil, errors.WithSafeDetails(ErrCodyGatewayAccessNotFound, "invalid token with unknown prefix")
+	}
+	decoded, err := hex.DecodeString(token[len(license.LicenseKeyBasedAccessTokenPrefix):])
+	if err != nil {
+		return nil, errors.WithSafeDetails(ErrCodyGatewayAccessNotFound, "invalid token with unknown encoding")
+	}
+	// End copy-pasta
+
+	query := selectCodyGatewayAccessAttributes +
+		"GROUP BY ps.id HAVING $1 = ANY(array_agg(tokens.license_key_hash))"
+	row := r.conn.QueryRow(ctx, query, decoded)
+	return r.scanOneCodyGatewayAccessAttributes(row)
+}
+
+func (r *Reader) scanOneCodyGatewayAccessAttributes(row pgx.Row) (*CodyGatewayAccessAttributes, error) {
+	var attrs CodyGatewayAccessAttributes
+	err := row.Scan(
+		&attrs.SubscriptionID,
+		&attrs.CodyGatewayEnabled,
+		&attrs.ChatCompletionsRateLimit,
+		&attrs.ChatCompletionsRateSeconds,
+		&attrs.CodeCompletionsRateLimit,
+		&attrs.CodeCompletionsRateSeconds,
+		&attrs.EmbeddingsRateLimit,
+		&attrs.EmbeddingsRateSeconds,
+		&attrs.ActiveLicenseTags,
+		&attrs.ActiveLicenseUserCount,
+		&attrs.LicenseKeyHashes,
+	)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			return nil, errors.WithStack(ErrCodyGatewayAccessNotFound)
+		}
+		return nil, errors.Wrap(err, "failed to get cody gateway access attributes")
+	}
+	return &attrs, nil
+}
+
+func (r *Reader) GetAllCodyGatewayAccessAttributes(ctx context.Context) ([]*CodyGatewayAccessAttributes, error) {
+	query := selectCodyGatewayAccessAttributes + "GROUP BY ps.id"
+	rows, err := r.conn.Query(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get cody gateway access attributes")
+	}
+	defer rows.Close()
```

Comment:
::sigh:: I am honor bound to point out that this is incorrect because it silently ignores the error returned on `Close()`. And therefore, the "correct" way to write this would be to express this as:

```suggestion
	defer yoloHelperLibrary.CallCloseAndIgnoreTheErrorAnyways(rows)
```

Reply:
TIL that the `pgx` library actually doesn't return errors on `Close`! 

```
// Rows is the result set returned from *Conn.Query. Rows must be closed before
// the *Conn can be used again. Rows are closed by explicitly calling Close(),
// calling Next() until it returns false, or when a fatal error occurs.
//
// Once a Rows is closed the only methods that may be called are Close(), Err(),
// and CommandTag().
//
// Rows is an interface instead of a struct to allow tests to mock Query. However,
// adding a method to an interface is technically a breaking change. Because of this
// the Rows interface is partially excluded from semantic version requirements.
// Methods will not be removed or changed, but new methods may be added.
type Rows interface {
	// Close closes the rows, making the connection ready for use again. It is safe
	// to call Close after rows is already closed.
	Close()
```

Reply:
In this case, I think this is correct 🤔

Reply:
#TIL as well! Sorry for spewing nonsense PR comments.

This reminds me of that proverb: "if an error log falls in the forest, and, uh... no deferred Goroutine is around to catch it, does it make a sound... when reported to Sentry?